### PR TITLE
Introduce the Root Circuit to the integration tests

### PR DESCRIPTION
**Pull request by ed255**
### Description

Introduce two new sets of tests in the Integration tests with the root circuit: one where the root circuit is verified with the MockProver and the other where a real proof is generated.  The input proof to the root circuit is the output of each subcircuit test we had before.

### Issue Link

Resolve https://github.com/privacy-scaling-explorations/zkevm-circuits/issues/1334
Resolve https://github.com/privacy-scaling-explorations/zkevm-circuits/issues/1460

### Type of change

- [x] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)

### Contents

Issue https://github.com/privacy-scaling-explorations/zkevm-circuits/issues/1335 has been resolved by assigning the fixed columns in the ExpTable and CopyTable in their `load` methods.  Previously they were either not assigned, or assigned dynamically.  This was not affecting the SuperCircuit or the Exp/Copy Circuit because in those cases, the assignment is done with a different method.  Only for SubCircuit tests where the SubCircuit requires the external table the `load` method is used.

### Rationale

Testing a Root Circuit (no mater if it's with the MockProver or with a real proof) requires a real proof of the underlying circuit.  This means that now for each SubCircuit we have 3 tests that require the SubCircuit proof:
- normal real prover test
- root mock prover test
- root real prover test

I implemented a cache to keep the SubCircuit proofs for each SubCircuit and block so that we have the possibility of running all the tests in a single run while reusing the SubCircuit proofs.

### How Has This Been Tested?

I have only tested the SubCircuits (minus SuperCircuit) + Root Circuit via MockProver.
- The machine I use has 64 GiB of memory, and that's not enough to generate a SuperCircuit proof
- I have as a pending task to try the Root Circuit real prover

But now we can run these tests under the github CI, so we can defer full testing after the PR is merged.

### CODE REVIEW

**Review by han0110 — APPROVED** [review]
LGTM! Nice work!

(for `SuperCircuit` it seems we'd need <code>2<sup>26</sup></code> size `RootCircuit` to aggregate from the experience of `zkevm-chain`)

**Comment by ed255** [comment]
> (for `SuperCircuit` it seems we'd need `2^26` size `RootCircuit` to aggregate from the experience of `zkevm-chain`)

Thanks for the hint! I've updated the tests in https://github.com/privacy-scaling-explorations/zkevm-circuits/pull/1471/commits/93dddd2c7d53a3b3c6b61616b910ef104fb010a7 to use degree=26 for Root Circuit with SuperCircuit and degree=24 for the other cases.

File: integration-tests/src/integration_test_circuits.rs [hunk 1]
```diff
@@ -34,13 +32,16 @@ use zkevm_circuits::{
...
     pi_circuit::TestPiCircuit,
+    root_circuit::{compile, Config, EvmTranscript, NativeLoader, RootCircuit, Shplonk},
     state_circuit::TestStateCircuit,
     super_circuit::SuperCircuit,
     tx_circuit::TestTxCircuit,
     util::SubCircuit,
     witness::{block_convert, Block},
 };
 
+type As<M> = Shplonk<M>;
```
Comment (CPerezz) [inline integration-tests/src/integration_test_circuits.rs:43]: This type alias is only used once in the entire codebase. I'd remove it TBH.

File: integration-tests/src/integration_test_circuits.rs [hunk 2]
```diff
@@ -282,31 +353,115 @@ impl<C: SubCircuit<Fr> + Circuit<Fr>> IntegrationTest<C> {
...
+                proof_cache.insert(proof_name, proof);
+            } else {
+                log::info!("circuit mock prover verification");
+                self.test_mock(&circuit, instance);
+            }
         }
     }
 }
 
+use zkevm_circuits::root_circuit::PoseidonTranscript;
```
Comment (CPerezz) [inline integration-tests/src/integration_test_circuits.rs:463]: Should we move this to the top of the file?

**Review by CPerezz — APPROVED** [review]
LGTM! 
Just 2 small suggestions. Feel free to ignore!

**Comment by ed255** [comment]
> LGTM! 
> Just 2 small suggestions. Feel free to ignore!

Thanks for the suggestions!  I've addressed them in https://github.com/privacy-scaling-explorations/zkevm-circuits/pull/1471/commits/957f6bd1afddc503e6b71b6b3c7e40ec8870766f

**Pull request merged**